tests/test_api_urls.py

```diff
@@ -66,31 +82,44 @@ def test_api_wrong_category(self):
 
     def test_api_labels_without_auth(self):
         '''API access to labels without auth returns JSON 200.'''
-        rv = self.app.get('/api/issues/labels', environ_base=headers)
-        json_body = json.loads(rv.data)
-        self.assertEqual(rv.status_code, 200)
-        self.assertEqual(rv.content_type, 'application/json')
+        with patch('webcompat.helpers.proxy_request') as github_data:
+            github_data.return_value = mock_api_response(
+                {'status_code': 200, 'content': '{}'})
+            rv = self.app.get('/api/issues/labels', environ_base=headers)
+            self.assertEqual(rv.status_code, 200)
+            self.assertEqual(rv.content_type, 'application/json')
 
     def test_api_comments_link_header_auth(self):
         '''API access to comments greater than 30 returns pagination in Link
         header of the response.'''
-        query_string = {'callback': 'foo'}
-        # Force a JSONP callback response with `query_string` because it gives
-        # us the header properties we want to test.
-        rv = self.app.get('/api/issues/398/comments',
-                          query_string=query_string, environ_base=headers)
-        self.assertTrue = all(x in rv.data for x in ['Link', 'rel', 'next',
-                                                     'last', 'page'])
-        self.assertEqual(rv.status_code, 200)
-        self.assertEqual(rv.content_type, 'application/json')
-        # API access to comments for an issue with < 30 does not return link a
-        # header in the response (until GitHub changes it....?)
-        rv = self.app.get('/api/issues/4/comments', query_string=query_string,
-                          environ_base=headers)
-        self.assertTrue = not all(x in rv.data for x in ['Link', 'rel', 'next',
+        with patch('webcompat.helpers.proxy_request') as github_data:
+            # Create mock data. One api_response for each call.
+            github_data.side_effect = [
+                mock_api_response({
+                    'status_code': 200,
+                    'content': '{"Link":[["https://api.github.com/repositories/17839063/issues/398/comments?callback=foo&page=2",{"rel":"next"}],["https://api.github.com/repositories/17839063/issues/398/comments?callback=foo&page=4",{"rel":"last"}]]}'  # nopep8
+                }),
```

Comment:
Here we might be able to do things slightly differently since we are now mocking our answers. This is what GitHub really returns:

```python
>>> import requests
>>> rv = requests.get('https://api.github.com/repositories/17839063/issues/398/comments')
>>> rv.links
{'last': {'url': 'https://api.github.com/repositories/17839063/issues/398/comments?page=4', 'rel': 'last'}, 'next': {'url': 'https://api.github.com/repositories/17839063/issues/398/comments?page=2', 'rel': 'next'}}
>>>
```

This is the part we are trying to test. webcompat returns this

```python
>>> rv = requests.get('http://localhost:5000/api/issues/398/comments')
>>> rv.links
{'last': {'url': '/api/issues/398/comments?page=4', 'rel': 'last'}, 'next': {'url': '/api/issues/398/comments?page=2', 'rel': 'next'}}
>>> 
```

You see we have rewritten the links.

So what we need to do now is to be able to fabricate headers containing the `Link`. Currently a GitHub answer is 

```python
>>> rv.headers
{'Access-Control-Allow-Origin': '*',
 'Access-Control-Expose-Headers': 'ETag, Link, X-GitHub-OTP, '
                                  'X-RateLimit-Limit, X-RateLimit-Remaining, '
                                  'X-RateLimit-Reset, X-OAuth-Scopes, '
                                  'X-Accepted-OAuth-Scopes, X-Poll-Interval',
 'Cache-Control': 'public, max-age=60, s-maxage=60',
 'Content-Encoding': 'gzip',
 'Content-Security-Policy': "default-src 'none'",
 'Content-Type': 'application/json; charset=utf-8',
 'Date': 'Tue, 08 Aug 2017 03:42:19 GMT',
 'ETag': 'W/"d5b26d0df7ccd5fa794d34177195129c"',
 'Link': '<https://api.github.com/repositories/17839063/issues/398/comments?page=2>; '
         'rel="next", '
         '<https://api.github.com/repositories/17839063/issues/398/comments?page=4>; '
         'rel="last"',
 'Server': 'GitHub.com',
 'Status': '200 OK',
 'Strict-Transport-Security': 'max-age=31536000; includeSubdomains; preload',
 'Transfer-Encoding': 'chunked',
 'Vary': 'Accept, Accept-Encoding',
 'X-Content-Type-Options': 'nosniff',
 'X-Frame-Options': 'deny',
 'X-GitHub-Media-Type': 'github.v3; format=json',
 'X-GitHub-Request-Id': 'DBA0:3BE9:E92B44:12683CD:5989331A',
 'X-RateLimit-Limit': '60',
 'X-RateLimit-Remaining': '59',
 'X-RateLimit-Reset': '1502167338',
 'X-Runtime-rack': '0.054068',
 'X-XSS-Protection': '1; mode=block'}
```

So basically we need to adjust the headers. But you can't do that without modifying a bit the magic function.

```python
def mock_api_response(response_config={}):
    '''Helper method to create a mock response from the Github API.'''
    api_response = MagicMock(spec=Response)
    api_response.content_type = 'application/json'
    api_response.headers = {
        'etag': 'thisisatest',
        'cache-control': 'thisisatest',
        'content-type': 'application/json'
    }
    for k, v in response_config.iteritems():
        if k == 'headers':
            v.update(api_response.headers)
        setattr(api_response, k, v)
    return api_response
```

We are updating the dictionary for headers. So we keep the base values and add new values.

And then the test becomes

```python
    def test_api_comments_link_header_auth(self):
        '''Link HTTP response header for comments page.'''
        with patch('webcompat.helpers.proxy_request') as github_data:
            github_data.side_effect = mock_api_response(
                {'status_code': 200,
                 'headers': {'Link': '<https://api.github.com/repositories/17839063/issues/398/comments?page=2>; rel="next", <https://api.github.com/repositories/17839063/issues/398/comments?page=4>; rel="last"'}})  # nopep8
            # Issues with more than 30 comments have a Link header.
            rv = self.app.get('/api/issues/398/comments', environ_base=headers)
            self.assertDictEqual(rv.links, {'last': {'url': '/api/issues/398/comments?page=4', 'rel': 'last'}, 'next': {'url': '/api/issues/398/comments?page=2', 'rel': 'next'}})  # nopep8
            self.assertEqual(rv.status_code, 200)
            self.assertEqual(rv.content_type, 'application/json')
            # Issues with less than 30 comments do NOT have a Link header.
            github_data.side_effect = mock_api_response({'status_code': 200})
            rv = self.app.get('/api/issues/4/comments', environ_base=headers)
            self.assertDictEqual(rv.links, {})
            self.assertEqual(rv.status_code, 200)
            self.assertEqual(rv.content_type, 'application/json')
```

Running the test now. Everything pass but one.

```
→ nosetests --verbose --nocapture tests/test_api_urls.py
Link HTTP response header for comments page. ... ERROR
API issue for a non existent number returns JSON 404. ... ok
API issue search with bad keywords returns JSON 404. ... ok
API access to labels without auth returns JSON 200. ... ok
API with wrong parameter returns JSON 404. ... ok
API setting labels without auth returns JSON 403 error code. ... ok
API access to user activity without auth returns JSON 401. ... ok
API with wrong category returns JSON 404. ... ok
API with wrong route returns JSON 404. ... ok
```

Ah! interesting!

Let's see the error message.

```
======================================================================
ERROR: Link HTTP response header for comments page.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/karl/code/webcompat.com/tests/test_api_urls.py", line 101, in test_api_comments_link_header_auth
    rv = self.app.get('/api/issues/398/comments', environ_base=headers)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/werkzeug/test.py", line 774, in get
    return self.open(*args, **kw)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/testing.py", line 127, in open
    follow_redirects=follow_redirects)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/werkzeug/test.py", line 742, in open
    response = self.run_wsgi_app(environ, buffered=buffered)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/werkzeug/test.py", line 659, in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/werkzeug/test.py", line 867, in run_wsgi_app
    app_rv = app(environ, start_response)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1994, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1985, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1540, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/karl/code/webcompat.com/webcompat/helpers.py", line 407, in wrapped_func
    return func(*args, **kwargs)
  File "/Users/karl/code/webcompat.com/webcompat/api/endpoints.py", line 226, in proxy_comments
    return api_request('get', path, params=params)
  File "/Users/karl/code/webcompat.com/webcompat/helpers.py", line 469, in api_request
    get_response_headers(resource))
  File "/Users/karl/code/webcompat.com/webcompat/helpers.py", line 199, in get_response_headers
    response.headers.get('link'))
  File "/Users/karl/code/webcompat.com/webcompat/helpers.py", line 330, in rewrite_and_sanitize_link
    return rewrite_links(sanitize_link(link_header))
  File "/Users/karl/code/webcompat.com/webcompat/helpers.py", line 286, in rewrite_links
    header_link_data = parse_link_header(link_header)
  File "/Users/karl/code/webcompat.com/webcompat/helpers.py", line 344, in parse_link_header
    uri_info, rel_info = link.split(';')
ValueError: need more than 1 value to unpack

----------------------------------------------------------------------
Ran 9 tests in 0.050s

FAILED (errors=1)
```

The test is helping us to discover a weakness in the parse_link_header function.